Sync from upstream: KEYBOARD.SYS, EGA.CPI, and related ones

FeralChild64 opened the pull request:
- synchronized binary blobs from upstream, moved them from *.h files to resources
- exposed mentioned blobs as files on Z: drive
- added more country definitions (and cleaned up the code here), so that all FreeDOS layout have not corresponding country code present
- fixed a small bug with AM/PM time display in shell commands - 12-hour format should display '1:23a', not '13:23a'

kcgen commented:
@Wengier , is it possible that your latest Z drive improvements mean that we can now bundle external Z: files as actual files?

I very much want us bundling files **_as files_**, and perhaps registering those files into Z: with a light-weight function call in the code - as opposed to storing binaries inside the source code.

Wengier commented:
@kcgen Yes, you can basically put any files in the drivez directory in the DOSBox-X directory. In the case of this PR the keyboard files can be put in the ``drivez\CPI`` directory, instead of storing them as binary blobs inside the source code.

FeralChild64 commented:
This would complicate deployment. Right now you can distribute just the binary, and it will just work. If the files are separate, than you need .SYS and .CPX files, or else the built-in KEYB command won't be functional. Besides, built-in KEYB is currently tied to the CPX/SYS files - it needs to know which SYS files contain keyboard layouts, which codepages can be found in which CPX files, etc.

Wengier commented:
@FeralChild64 Yes, I am aware that this will likely have some benefits as you said, although there appears to be a previous discussion which encourages putting actual files instead. I think it may be discussed for this specific case.

FeralChild64 commented:
Well, as you decide. The reason behind this pull request is that I'm slowly cooking (when I'm not in the mood to do some harder work) a locale upgrade to match the current FreeDOS level - for this I will need up to date keyboard/font definitions. It's not urgent, as there is quite a lot to update, see https://gitlab.com/FreeDOS/base/keyb_lay/-/blob/master/DOC/KEYB/LAYOUTS/LAYOUTS.TXT

kcgen commented:
We know there are lots of binary files that make sense to ship alongside the binary: shaders, translation files, keyb files, TTF fonts, (perhaps) ROMs, (perhaps) a tiny SF2, and so on. 

Staging already needs critical DLLs just to get off the ground, and I think it's reasonable that we be able to ship other files critical to proper operation alongside our executable.

Modern games and applications expect to have a local filesystem to support their own data files, and Staging should be the same. 

Up till now this has been a bit challenging because discovering the executable path is not trivial in a cross-platform way.  Staging's `support.h` includes the `GetExecutablePath()` function:

``` c++
const std_fs::path &GetExecutablePath();
```

And you can work with that return value to directly construct paths to data files. A rough example:

``` c++
const auto layouts_path = GetExecutablePath() / "layouts";
const auto ega_cpx = layouts_path / "EGA.CPX";
const auto ega2_cpx = layouts_path / "EGA2.CPX";
// ega2_cpx can be used as a std::string or use `.c_str()` to get a const char *
```

Then a layout file read functions can read it into a mem buffer.

FeralChild64 commented:
I would rather prefer to have something like `const std_fs::path &GetResourcePath();`, as on Linux resources are typically located somewhere within `/usr/share`, and I have no idea about MacOS filesystem layout. I'll look into this after the weekend.

kcgen commented:
> I would rather prefer to have something like `const std_fs::path &GetResourcePath();`, as on Linux resources are typically located somewhere within `/usr/share`, and I have no idea about MacOS filesystem layout. I'll look into this after the weekend.

I like that a lot @FeralChild64!

I think we can make the function take in a resource **sub-directory** and  **file**, and it will be aware of the platform-specific Staging resource path(s) and their search order, and will hand back the leading hit for the requested file.

`cont auto ega_cpx = GetResource(layouts / "EGA.CPX");`

So for example, it will first check in our bundled resource tree packaged with the executable (for users who only unzip and run), then it will look in their home directory for Staging's resource areas (XDG paths on Linux, Resource path on macOS), and finally (on macOS and Linux): `/usr/local/share/dosbox-staging/resources/...` and `/usr/share/dosbox-staging/resources/...`, for users who install via package manager.

kcgen commented:
@FeralChild64 , if you can can rebase against `main`, it now has the improved resource handling you suggested.

To add your new resources:

1. Create them in a new directory under `contrib/resources`, I'll call them `layouts/`:

![2022-03-26_09-27](https://user-images.githubusercontent.com/1557255/160248582-093c2eb6-c069-40ad-be44-823dcf9c0c7f.png)

2. The directory can contain deliverables and non-deliverables (I added couple examples).
3. Add a `meson.build` file inside with a "resources" list of just the deliverables:

    ``` python
    resources = [
      'ega.cpx',
      'ega2.cpx',
      'ega3.cpx',
    ]
    ```
4. Add `layouts` to the list of resource_dirs at the top of `contrib/resources/meson.build`:

    ``` python
    resource_dirs = [
      'translations',
      'layouts',
    ]
    ```
    
5. In your code, use two ways to get std_fs::path objects to your files:

    ``` c++
    std_fs::path GetResource(const std_fs::path &name);
    std_fs::path GetResource(const std_fs::path &subdir, const std_fs::path &name);
    // examples
    const auto layout_file = GetResource("layouts", "ega.cpx");
    const auto layout_file = GetResource("layouts/ega.cpx");
    ```

The resource files are bundled and shipped in all packages.

FeralChild64 commented:
OK, I'll let you know when it's time to restart the review.

kcgen commented on src/dos/dos.cpp:
```diff
@@ -49,114 +49,143 @@ void DOS_SetError(Bit16u code) {
-        COUNTRY::Latin_America,
-        COUNTRY::Australia,
-};
+static constexpr Bit8u DATE_MDY = 0;
+static constexpr Bit8u DATE_DMY = 1;
+static constexpr Bit8u DATE_YMD = 2;
```
This is looking top-notch @FeralChild64 !

super minor comment: Staging and -X have both deprecated DOSBox's custom types (`Bitu8u`, `Bitu`, etc.. ) for the usual C99 `#include <stdint>` types.. `uint8_t`, etc.

FeralChild64 commented:
OK, you may now resume the review. `Bit8u` is now `uint8_t`, FreeDOS blobs are now resources (I have added blobs, license, documentation, and zipped source if available - this should cover the GPL v2 requirements). I have also noticed a small bug with 12-hour time display, this is also fixed now.

Would it be possible to squish commits during the merge? There is some junk in older commits, which is in fact not necessary.

shermp commented:
@FeralChild64 feel free to rewrite this branches history as you feel fit, and force push. You'll notice we tend to do that quite often around these parts.

Wengier commented:
@FeralChild64 You probably want to load ``BLOB_EGA18_CPX`` instead of ``BLOB_EGA10_CPX`` in line 876 of src/dos_keyboard_layout.cpp. Also, you may want to load ``keybrd4.sys`` too in line 301 and 674 of the same file. Moreover, have a look at "auto" detection in the read_codepage_file function, which may be updated as well to match the updated files.

kcgen commented:
Squashed and rebased your branch, and created it here in this repo. 

Your resources load function is a nice generic helper, so I moved that into support.cpp with a couple small adjustments (can specify a resource being mandatory or optional). 

Feel free to take this branch over now; and start pushing changes to address @Wengier 's suggestions.

kcgen commented:
@FeralChild64 , suggest we carry on reviewing this PR using the local `fc/keyb-updates-2` (renamed just to avoid ambiguity) branch here.
It's been rebased against main, and squashed to eliminate the prior source-blob commits, and is passing all builds and unit tests.

(ok.. closing, please jump over to: https://github.com/dosbox-staging/dosbox-staging/compare/fc/keyb-updates-2?expand=1 )
